controlplane/kubeadm/internal/kubeadm_config_map.go

```diff
@@ -196,6 +202,83 @@ func (k *kubeadmConfig) UpdateCoreDNSImageInfo(repository, tag string) error {
 	return nil
 }
 
+// UpdateAPIServer sets the api server configuration to values set in `apiServer` in kubeadm config map.
+func (k *kubeadmConfig) UpdateAPIServer(apiServer kubeadmv1.APIServer) (bool, error) {
+	changed, err := k.updateClusterConfiguration(apiServer, apiServerKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update api server configuration in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// UpdateControllerManager sets the controller manager configuration to values set in `controllerManager` in kubeadm config map.
+func (k *kubeadmConfig) UpdateControllerManager(controllerManager kubeadmv1.ControlPlaneComponent) (bool, error) {
+	changed, err := k.updateClusterConfiguration(controllerManager, controllerManagerKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update controller manager configuration in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// UpdateScheduler sets the scheduler configuration to values set in `scheduler` in kubeadm config map.
+func (k *kubeadmConfig) UpdateScheduler(scheduler kubeadmv1.ControlPlaneComponent) (bool, error) {
+	changed, err := k.updateClusterConfiguration(scheduler, schedulerKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update scheduler configuration in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// UpdateFeatureGates sets the feature gates to values set in `featureGates` in kubeadm config map.
+func (k *kubeadmConfig) UpdateFeatureGates(featureGates map[string]bool) (bool, error) {
+	changed, err := k.updateClusterConfiguration(featureGates, featureGatesKey)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to update feature gates in kubeadm config map")
+	}
+	return changed, nil
+}
+
+// updateClusterConfiguration is a generic method to update any kubeadm ClusterConfiguration spec with custom types in the specified path.
+func (k *kubeadmConfig) updateClusterConfiguration(config interface{}, path ...string) (bool, error) {
+	data, ok := k.ConfigMap.Data[clusterConfigurationKey]
+	if !ok {
+		return false, errors.Errorf("unable to find %q in kubeadm ConfigMap", clusterConfigurationKey)
+	}
+
+	configuration, err := yamlToUnstructured([]byte(data))
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to decode kubeadm ConfigMap's %q to Unstructured object", clusterConfigurationKey)
+	}
+
+	currentConfig, _, err := unstructured.NestedFieldCopy(configuration.UnstructuredContent(), path...)
+	if err != nil {
```

Comment:
Wondering if we should use zero instead of throwing an error in this case

Reply:
I don't fully understand the comment. Do you have any particular scenario in mind?

Reply:
re-reading this I think this is ok, because assuming zero in case of errors could lead to false assumptions